### ISSUE TITLE
Report missing decoders and other audio decoder errors

### ISSUE THREAD

**Issue by ferjm** [servo/media#125]
We should be listening for pipeline bus error messages and firing the audio decoder error callback when needed.

```
philn> ferjm: there you go :)
12:27:54 travis needs ogg/vorbis libs
12:28:21 <ferjm> Missing decoder: Ogg (audio/ogg)
12:28:33 cool
12:30:18 we shouldn't be freezing there though :\
12:33:03 philn: I guess decodebin is throwing errors for missing decoders, right?
12:34:30 ⇐ BK1603 quit (uid316735@moz-jgv72q.stonehaven.irccloud.com) Quit: Connection closed for inactivity
12:36:35 <philn> yeah
12:38:58 you should be able to reproduce the error locally by moving out libgstogg.so
12:39:16 <@Manishearth> philn: thanks :)
12:39:36 <ferjm> philn: ok, cool. Are these errors sent through as a pipeline bus message or is there an error callback I can handle in decodebin?
12:39:59 <@Manishearth> ferjm: in the example we wait for it to respond, we should .unwrap() the .recv() call
12:40:02 that's why it hangs
12:40:03 <philn> if you use gst-player you can connect to the error signal iirc
12:40:08 <@Manishearth> we get stuck on .recv()
12:40:26 <philn> otherwise yes, an error message is sent over the bus
12:41:10 <@Manishearth> philn: its a bug in the test's error handling
12:41:16 → rohitpaulk23833258 joined (loungeuser@moz-emn.5fb.201.103.IP)
12:41:22 <philn> ah, ok!
12:42:15 <ferjm> philn: ok, we are not using gst-player in this case
12:42:43 Manishearth:  https://github.com/servo/media/blob/master/examples/audio_decoder.rs#L50
12:43:17 <@Manishearth> ferjm: oh hm
12:43:18 <ferjm> we should be firing the error callback for the missing decoder and send a message through the channel there https://github.com/servo/media/blob/master/examples/audio_decoder.rs#L34
12:43:23 <@Manishearth> yeah
12:43:34 it's an eror handler so it doesn't panic, my bad
12:43:34 <ferjm> but we need to get the error from the pipeline bus first 
```

**Comment by sdroege** [comment]
What kind of errors would you like the player to export? Currently the error enum there is just a stub. Having separate errors for missing decoders, file not found, etc. would definitely be something I'd like to have in there but I didn't get to design anything yet. Errors are hard :)

If you can come up with a proposal what you would like to have that would be very helpful.

**Comment by sdroege** [comment]
Sure

**Comment by ferjm** [comment]
Yeah, we are not using gst-player in this case. But we will be using it for `<video>` soon :), so we'll need these errors in place.

**Comment by ceyusa** [comment]
And in this point we could ask ourselves if we just want to support a certain set of codecs or all those available by the user.

**Comment by philn** [comment]
I would advise to inspect the registry and advertise only the supported codecs.

**Comment by ferjm** [comment]
This was fixed by #139
